File: Parcial 2/Actividad/Actividad.py
class Animal:
    def __init__(self, name):
        self.name = name
        self._age = 0
        self._color = None  # Atributo privado color

    def speak(self):
        raise NotImplementedError("Subclass must implement abstract method")

    def comer(self):
        return f"{self.name} is eating."

    def setColor(self, color):
        self._color = color

    def getColor(self):
        return self._color

    def set_age(self, age):
        self._age = age

    def get_age(self):
        return self._age

class Dog(Animal):
    def speak(self):
        return f"{self.name} says Woof!"
    
    def comer(self):
        return f"{self.name} is eating dog food."

class Cat(Animal):
    def speak(self):
        return f"{self.name} says Meow!"
    
    def comer(self):
        return f"{self.name} is eating cat food."

# Nueva clase derivada: Fish
class Fish(Animal):
    def speak(self):
        return f"{self.name} doesn't speak. It swims silently."
    
    def comer(self):
        return f"{self.name} is eating fish food."

# Nueva clase derivada: Bird
class Bird(Animal):
    def speak(self):
        return f"{self.name} says Chirp!"
    
    def comer(self):
        return f"{self.name} is eating bird seed."

# Ejemplo de uso
dog = Dog("Rex")
cat = Cat("Whiskers")
fish = Fish("Goldie")
bird = Bird("Tweety")

# Establecer colores
dog.setColor("Brown")
cat.setColor("Black")
fish.setColor("Gold")
bird.setColor("Yellow")

print(dog.speak())  # Rex says Woof!
print(cat.speak())  # Whiskers says Meow!
print(fish.speak())  # Goldie doesn't speak. It swims silently.
print(bird.speak())  # Tweety says Chirp!

# Métodos comer
print(dog.comer())  # Rex is eating dog food.
print(cat.comer())  # Whiskers is eating cat food.
print(fish.comer())  # Goldie is eating fish food.
print(bird.comer())  # Tweety is eating bird seed.

# Métodos de edad y color
dog.set_age(3)
cat.set_age(5)
print(dog.get_age())  # 3
print(cat.get_age())  # 5

print(dog.getColor())  # Brown
print(cat.getColor())  # Black
print(fish.getColor())  # Gold
print(bird.getColor())  # Yellow

File: Parcial 2/Actividad/Actividad.java
class Animal {
    private String name;
    private int _age;
    private String _color;  // Atributo privado color

    public Animal(String name) {
        this.name = name;
        this._age = 0;
        this._color = null;
    }

    public void speak() {
        throw new UnsupportedOperationException("Subclass must implement abstract method");
    }

    public String comer() {
        return name + " is eating.";
    }

    public void setColor(String color) {
        this._color = color;
    }

    public String getColor() {
        return this._color;
    }

    public void setAge(int age) {
        this._age = age;
    }

    public int getAge() {
        return this._age;
    }

    public String getName() {
        return this.name;
    }
}

class Dog extends Animal {
    public Dog(String name) {
        super(name);
    }

    @Override
    public void speak() {
        System.out.println(getName() + " says Woof!");
    }

    @Override
    public String comer() {
        return getName() + " is eating dog food.";
    }
}

class Cat extends Animal {
    public Cat(String name) {
        super(name);
    }

    @Override
    public void speak() {
        System.out.println(getName() + " says Meow!");
    }

    @Override
    public String comer() {
        return getName() + " is eating cat food.";
    }
}

// Nueva clase derivada: Fish
class Fish extends Animal {
    public Fish(String name) {
        super(name);
    }

    @Override
    public void speak() {
        System.out.println(getName() + " doesn't speak. It swims silently.");
    }

    @Override
    public String comer() {
        return getName() + " is eating fish food.";
    }
}

// Nueva clase derivada: Bird
class Bird extends Animal {
    public Bird(String name) {
        super(name);
    }

    @Override
    public void speak() {
        System.out.println(getName() + " says Chirp!");
    }

    @Override
    public String comer() {
        return getName() + " is eating bird seed.";
    }
}

public class Main {
    public static void main(String[] args) {
        Dog dog = new Dog("Rex");
        Cat cat = new Cat("Whiskers");
        Fish fish = new Fish("Goldie");
        Bird bird = new Bird("Tweety");

        // Establecer colores
        dog.setColor("Brown");
        cat.setColor("Black");
        fish.setColor("Gold");
        bird.setColor("Yellow");

        System.out.println(dog.getName() + " says: " + dog.comer());
        System.out.println(cat.getName() + " says: " + cat.comer());
        System.out.println(fish.getName() + " says: " + fish.comer());
        System.out.println(bird.getName() + " says: " + bird.comer());

        // Métodos de edad y color
        dog.setAge(3);
        cat.setAge(5);
        System.out.println(dog.getName() + " age: " + dog.getAge());
        System.out.println(cat.getName() + " age: " + cat.getAge());

        System.out.println(dog.getName() + " color: " + dog.getColor());
        System.out.println(cat.getName() + " color: " + cat.getColor());
        System.out.println(fish.getName() + " color: " + fish.getColor());
        System.out.println(bird.getName() + " color: " + bird.getColor());
    }
}
